### ISSUE TITLE
Show titan ACL options in WCF admin dashboard

### ISSUE THREAD

**Issue by MichaelJamesParsons** [unkso/titan-app#79]
## Description
When updating a user's permissions in WCF, admins should be able to assign/revoke titan's ACL options for users and user groups.

## Requirements/Expected Behavior
Update WCF to expose titan's ACL options.

**Comment by ELD** [comment]
@MichaelJamesParsons I investigated this a little yesterday and tonight. Here are my findings.

**TL;DR: I don't think it'll be too difficult if we want to attempt this now. It's probably in our best interesting to do so.**

I was looking into wiring up our ACL options to WBB's ACP. The short of it is that we operated under an incorrect assumption, however, we can remedy this situation.

The `wcf1_acl_option` table is what specifies permissions that can be assigned to users or usergroups. Typically, these are assigned to usergroups, not to individual users. In the case of forum permissions, you can specify one-off permissions, but it's much more the convention to assign them to groups and assign those groups to users.

In order to expose these in the ACP, we need to add the options to XML templates. A good example of this is the `unkso/base` repo with the `userGroupOption.xml` file. Combined with the `en.xml` file, this sets the name of the options that are exposed in the ACP.

What I propose is that we create a Titan ACL package for WBB that adds those options that we can assign to usergroups. From there, we have to reconfigure how we fetch ACL, but I was thinking that we could get to individual ACL permissions by means of the individual options values. 

This would be similar to how WBB does it. An example of this is in `unkso/awards` for `AwardListPage.class.php`. It specifies permissions and the user object is checked to see if they have the permissions. I'm not sure if we can reliably cache the ACL permissions client-side to ensure appropriate authorization. In fact, I'd argue that we all authorization and authentication should happen server side (auth current is done this way). This would be a more secure authentication/authorization measure.

**Comment by MichaelJamesParsons** [comment]
> From there, we have to reconfigure how we fetch ACL, but I was thinking that we could get to individual ACL permissions by means of the individual options values.

Titan derives a user's ACL permissions through options that have been directly assigned to the individual as well as any group they're a member of. Assigning options to groups instead of individuals should work out of the box with what we have now.

> I'm not sure if we can reliably cache the ACL permissions client-side to ensure appropriate authorization.
> ...
> I'd argue that we all authorization and authentication should happen server side.

I agree with both of these statements. ACL options are cached locally on the front end so the UI can know which components to hide when the authenticated user doesn't hold adequate permissions. The sole purpose of this approach is to optimize the user experience without checking the server for ACL options on every page load. Earlier, you mentioned implementing a token refresh mechanism. That refresh might be a good opportunity to reload ACL options as well.

These client-side checks are not intended to take the place of server side ACL authorization. As of now, server side checks are handled inline with the request handler. Moving this logic to a request guard would be a much cleaner approach.

**Comment by ELD** [comment]
> Titan derives a user's ACL permissions through options that have been directly assigned to the individual as well as any group they're a member of. Assigning options to groups instead of individuals should work out of the box with what we have now.

Great, then this should be really straight forward. I'll start working on the plugin tonight and check it into this repo.

> I agree with both of these statements. ACL options are cached locally on the front end so the UI can know which components to hide when the authenticated user doesn't hold adequate permissions. The sole purpose of this approach is to optimize the user experience without checking the server for ACL options on every page load. Earlier, you mentioned implementing a token refresh mechanism. That refresh might be a good opportunity to reload ACL options as well.
>
> These client-side checks are not intended to take the place of server side ACL authorization. As of now, server side checks are handled inline with the request handler. Moving this logic to a request guard would be a much cleaner approach.

The caching makes sense, we just need to build a way, long term, to invalidate the cache so that people can't retain permissions they shouldn't have after an event (i.e. leaving the clan, getting A-15'd, etc). The refresh mechanism, I think, would address those concerns. Definitely something we should look into in the near-term.

Moving those checks to a request guard would be a much cleaner and Rocket-idiomatic approach. That feels like a technical chore, since it's a refactor. Feel free to create an issue and schedule it accordingly.
